### PR TITLE
[WB-7886] Add CatBoost Integration

### PR DESCRIPTION
**Pull request by ayulockin**
Fixes WB-7886

Description
-----------
The PR adds a simple `WandbCallback` for [CatBoost](https://catboost.ai/). 

The PR currently enables:
- logging iteration 
- logging train and validation metrics

Testing
-------
`yea` test and manually tested.

### CODE REVIEW

File: wandb/integration/catboost/catboost.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    Passing `WandbCallback` to CatBoost will:
+    - log training and validation metrics at every `metric_period`
+    - log iteration at every `metric_period`
+
+    Example:
+        ```
+        train_pool = Pool(train[features], label=train['label'], cat_features=cat_features)
+        test_pool = Pool(test[features], label=test['label'], cat_features=cat_features)
+
+        model = CatBoostRegressor(iterations=100,
```
Comment (morganmcg1) [inline wandb/integration/catboost/catboost.py:24]: nit, maybe move the `iterations` argument to a new line

File: wandb/integration/catboost/catboost.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            loss_function='Cox',
+            eval_metric='Cox',
+        )
+
+        model.fit(train_pool,
```
Comment (morganmcg1) [inline wandb/integration/catboost/catboost.py:29]: formatting, lets keep this consistent with the open bracket / close bracket styling in `model = CatBoostRegressor...`

**Review by morganmcg1 — COMMENTED** [review]
Look good so far, the new XGB callback code is included here too, that should be kept in a separate PR.

There is a `log_function` function to be added too right?

File: wandb/integration/catboost/catboost.py (hunk 1 above)
Comment (dmitryduev) [reply wandb/integration/catboost/catboost.py:24]: 👍🏻 In general, please re-format the example code. I'd copy it into a dummy file, the run `tox -e format -- dummy.py` and copy back into docstrings.

File: wandb/integration/catboost/catboost.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                  eval_set=test_pool,
+                  callbacks=[WandbCallback()])
+        ```
+    """
+
+    def __init__(self, metric_period: int = 1):
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before WandbCallback()")
```
Comment (dmitryduev) [inline wandb/integration/catboost/catboost.py:37]: Could you please add a test case for this? Add a try/except clause to your functional test? We're trying to please codecov as much as possible...

File: wandb/integration/catboost/catboost.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        self.metric_period = metric_period
+
+    def after_iteration(self, info):
```
Comment (dmitryduev) [inline wandb/integration/catboost/catboost.py:41]: type-annotate please

File: functional_tests/catboost/catboost.yea [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+        - catboost>=0.26
+        
+assert:
+  - :wandb:runs_len: 1
+  - :wandb:runs[0][summary][iteration@metric-period-1]: 10
+  - :op:>:
+    - :wandb:runs[0][summary][learn-MultiClass]
+    - 0.0
+  - :wandb:runs[0][exitcode]: 0
+
```
Comment (dmitryduev) [inline functional_tests/catboost/catboost.yea:17]: formatting

File: functional_tests/catboost/test_catboost.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python
+"""Test new catboost integration."""
```
Comment (dmitryduev) [inline functional_tests/catboost/test_catboost.py:2]: It will become not-so-new eventually... :)
```suggestion
"""Test CatBoost integration."""
```

**Review by dmitryduev — COMMENTED** [review]
Thanks @ayulockin! Mostly LGTM, a few minor comments.
Please rm the xgboost stuff from this PR. Also, @raubitsj, I saw you were gonna add telemetry changes to the xgboost PR, would you mind doing the same here please?

**Comment by dmitryduev** [comment]
@raubitsj looking at `wandb/proto/wandb_telemetry.proto`, seems like the stuff to enable telemetry for `catboost` is already there, is that right?

File: wandb/integration/catboost/catboost.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
```
Comment (dmitryduev) [inline wandb/integration/catboost/catboost.py:1]: We're py3 only, so no need to specify this.
```suggestion
```

File: wandb/integration/catboost/catboost.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+        model = CatBoostRegressor(
+            iterations=100,
+            loss_function='Cox',
+            eval_metric='Cox',
+        )
+
+        model.fit(train_pool,
+                  eval_set=test_pool,
+                  callbacks=[WandbCallback()])
```
Comment (dmitryduev) [inline wandb/integration/catboost/catboost.py:36]:
```suggestion
        model.fit(
            train_pool,
            eval_set=test_pool,
            callbacks=[WandbCallback()]
        )
```

**Review by dmitryduev — APPROVED** [review]
LGTM!

**Comment by ayulockin** [comment]
Thanks for making the fixes. LGTM too.

File: tests/test_library_public.py [hunk 9]
```diff
@@ -121,9 +124,9 @@
     "wandb_sdk",
     "wandb_torch",
     "xgboost",
-    "sync",
-    "sweeps",
     "cli",
+    # typing
+    "Optional",
```
Comment (kptkin) [inline tests/test_library_public.py:129]: I think we should avoid this?

Comment (dmitryduev) [reply tests/test_library_public.py:129]: Why? This is for `wandb.run` init in `wandb.__init__`. I would eventually like to type-annotate all of the latter, and in the case of wandb.run it has to include Optional.

Comment (kptkin) [reply tests/test_library_public.py:129]: ah just, since this test check for public functions to avoid exposing internal things, maybe it should be in the excluded category (for the tests)?

File: mypy.ini [hunk 10]
```diff
@@ -34,6 +34,19 @@ disallow_untyped_decorators = True
...
 
+[mypy-wandb.integration.catboost.*]
+ignore_errors = False
+disallow_incomplete_defs = True
+disallow_untyped_defs = True
+warn_unused_ignores = True
+warn_return_any = True
+warn_unreachable = True
+check_untyped_defs = True
+# disallow_untyped_calls = True
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply mypy.ini:45]: just tried and mypy gets unhappy bc of some other untyped stuff, but only unrelated to this pr, so let's leave it as is for now.

File: wandb/integration/catboost/catboost.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        if wandb.run is None:
+            raise wandb.Error("You must call `wandb.init()` before `WandbCallback()`")
+
+        self.metric_period: int = metric_period
+
+    def after_iteration(self, info: SimpleNamespace) -> bool:
+        if info.iteration % self.metric_period == 0:
+            for data, metric in info.metrics.items():
+                for metric_name, log in metric.items():
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
```
Comment (kptkin) [inline wandb/integration/catboost/catboost.py:52]:
```suggestion
                    wandb.run._log({f"{data}-{metric_name}": log[-1]}, commit=False)
```
so maybe we can create an internal function in the run class:
I'm adding in my PR:
https://github.com/wandb/client/pull/3137/files#diff-b63db14fd18e0e4d240ec68e625f5420067376cb574e03f383ca7a29e7bfce4bR1134

File: wandb/integration/catboost/catboost.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+            wandb.log({f"iteration@metric-period-{self.metric_period}": info.iteration})
```
Comment (kptkin) [inline wandb/integration/catboost/catboost.py:54]: same as above

File: tests/test_library_public.py (hunk 9 above)
Comment (dmitryduev) [reply tests/test_library_public.py:129]: as in, add another set of excluded symbols?

File: wandb/integration/catboost/catboost.py (hunk 11 above)
Comment (dmitryduev) [reply wandb/integration/catboost/catboost.py:52]: Should I commit this now?

File: tests/test_library_public.py (hunk 9 above)
Comment (kptkin) [reply tests/test_library_public.py:129]: yep... like here (we exclude all internal symbols)
https://github.com/wandb/client/blob/master/tests/test_library_public.py#L132

File: wandb/integration/catboost/catboost.py (hunk 11 above)
Comment (kptkin) [reply wandb/integration/catboost/catboost.py:52]: add a todo... sorry that I used the suggest...
it won't work (we don't have internal function)

File: wandb/sdk/data_types.py [hunk 13]
```diff
@@ -80,9 +80,9 @@ def _get_max_cli_version() -> Union[str, None]:
 
 
 def _is_offline() -> bool:
-    return (
-        wandb.run is not None and wandb.run._settings.mode == "offline"  # type: ignore
-    ) or str(wandb.setup().settings.mode) == "offline"
+    return (wandb.run is not None and wandb.run._settings.mode == "offline") or str(
```
Comment (kptkin) [inline wandb/sdk/data_types.py:83]: why is this offline different from the settings._offline :/ (I know it's not you... just weird that we have overloaded definitions

Comment (dmitryduev) [reply wandb/sdk/data_types.py:83]: will fix :)

File: wandb/integration/catboost/catboost.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+    model_name = f"model_{wandb.run.id}"
+    # save the model in the default `cbm` format
+    model_path = Path(wandb.run.dir) / "model"
+
+    model.save_model(model_path)
+
+    model_artifact = wandb.Artifact(name=model_name, type="model")
+    model_artifact.add_file(str(model_path))
+    wandb.log_artifact(model_artifact, aliases=aliases)
```
Comment (kptkin) [inline wandb/integration/catboost/catboost.py:78]: probably here we also should use the internal log_artifact: `wandb.run._log_artifact`

File: wandb/sdk/wandb_artifacts.py [hunk 15]
```diff
@@ -615,7 +615,7 @@ def save(
                             tel.feature.artifact_incremental = True
                     run.log_artifact(self)
             else:
-                wandb.run.log_artifact(self)  # type: ignore
+                wandb.run.log_artifact(self)
```
Comment (kptkin) [inline wandb/sdk/wandb_artifacts.py:618]: should we use `_log_artifact` here as well?

File: wandb/integration/catboost/catboost.py (hunk 14 above)
Comment (dmitryduev) [reply wandb/integration/catboost/catboost.py:78]: also todo?

File: wandb/sdk/wandb_artifacts.py (hunk 15 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_artifacts.py:618]: it was just for fixing mypy errors, I'd leave it for another PR

File: wandb/integration/catboost/catboost.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+    # log best score and iteration
+    wandb.run.summary["best_iteration"] = model.get_best_iteration()
+    wandb.run.summary["best_score"] = model.get_best_score()
+
+    # log model
+    if save_model_checkpoint:
+        if params["use_best_model"]:
+            _checkpoint_artifact(model, aliases=["best"])
+        else:
+            _checkpoint_artifact(model, aliases=["last"])
```
Comment (kptkin) [inline wandb/integration/catboost/catboost.py:172]:
```suggestion
        aliases = ["best"] if params["use_best_model"] else ["last"]
        _checkpoint_artifact(model, aliases=)
```
nit: it's just my stylistic preference

File: wandb/integration/catboost/catboost.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+                for metric_name, log in metric.items():
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
+
+            wandb.log({f"iteration@metric-period-{self.metric_period}": info.iteration})
+
+        return True
+
+
+def _checkpoint_artifact(
+    model: Union[CatBoostClassifier, CatBoostRegressor], aliases: list
```
Comment (kptkin) [inline wandb/integration/catboost/catboost.py:60]: list of what?

Comment (dmitryduev) [reply wandb/integration/catboost/catboost.py:60]: Fixed

File: functional_tests/catboost/test_catboost.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
+"""Test CatBoost integration."""
+
+import catboost
+from catboost import datasets
```
Comment (kptkin) [inline functional_tests/catboost/test_catboost.py:5]:
```suggestion
from catboost import CatBoostClassifier, datasets, Pool
```
what do you think?

File: wandb/integration/catboost/catboost.py (hunk 14 above)
Comment (kptkin) [reply wandb/integration/catboost/catboost.py:80]: no... we actually have this function :)))
https://github.com/wandb/client/blob/55b885c14018677b916e006ca9dbc9814db99d7c/wandb/sdk/wandb_run.py#L2165

File: functional_tests/catboost/test_catboost.py (hunk 18 above)
Comment (dmitryduev) [reply functional_tests/catboost/test_catboost.py:5]: 👍🏻

File: wandb/sdk/wandb_artifacts.py (hunk 15 above)
Comment (kptkin) [reply wandb/sdk/wandb_artifacts.py:618]: sounds good

File: wandb/integration/catboost/catboost.py (hunk 14 above)
Comment (dmitryduev) [reply wandb/integration/catboost/catboost.py:79]: Oops. What's your preference? I don't care :) Just generally prefer using public interfaces ;)

Comment (kptkin) [reply wandb/integration/catboost/catboost.py:79]: hmmm... even for integration? I'm not sure... 
Just raising this as @raubitsj suggested using internal interface in my PR

File: wandb/catboost/__init__.py [hunk 19]
```diff
@@ -0,0 +1,10 @@
+"""
+Compatibility catboost module.
+
+In the future use:
```
Comment (kptkin) [inline wandb/catboost/__init__.py:4]: does it mean we are going to deprecate this interface?
should we add a deprecation message here if we do?

Comment (dmitryduev) [reply wandb/catboost/__init__.py:4]: I suggest we add deprecation messages to all our compatibility modules in a separate PR, otherwise things will be inconsistent.

Comment (kptkin) [reply wandb/catboost/__init__.py:4]: sure... let's track this in an issue?

Comment (dmitryduev) [reply wandb/catboost/__init__.py:4]: Done!

File: wandb/integration/catboost/catboost.py (hunk 14 above)
Comment (raubitsj) [reply wandb/integration/catboost/catboost.py]: In this case there isnt a reason now to prefer _log_artifact over log_artifact. 
If it was using some weird feature of log_artifact and we wanted capture that in a telemetry object then using the internal one would be better.. but this looks pretty standard, it is fine how it is.

**Comment by raubitsj** [comment]
> Thanks @ayulockin! Mostly LGTM, a few minor comments.
> 
> * Tested it out using their basic tutorial https://github.com/catboost/tutorials/blob/master/python_tutorial.ipynb -- all works well including things like logging custom metrics)
> * Please rm the xgboost stuff from this PR.
> * Also, @raubitsj, I saw you were gonna add telemetry changes to the xgboost PR, would you mind doing the same here please?

This is the type of changes needed for telemetry:
https://github.com/wandb/client/pull/3024/commits/80afe3f57e01424686104a1bff3ee9bcf802097f

Let me know if you need help, hopefully this is a clean change to model after.

**Comment by raubitsj** [comment]
> 



> @raubitsj looking at `wandb/proto/wandb_telemetry.proto`, seems like the stuff to enable telemetry for `catboost` is already there, is that right?

Responded at: https://github.com/wandb/client/pull/2975#issuecomment-1042502819

**Comment by dmitryduev** [comment]
@raubitsj: I added feature usage tracking to telemetry + testing that with `yea` as you requested. Passes locally and the CI is green, so merging. 
Many thanks for this once again, @ayulockin!!

**Comment by ayulockin** [comment]
Thanks @dmitryduev for shipping it.
